File: 数据结构部分/1-100/_24.cpp
#include "normal.h"

/*
给定一个链表，两两交换其中相邻的节点，并返回交换后的链表。

你不能只是单纯的改变节点内部的值，而是需要实际的进行节点交换。
*/


//Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if (head == NULL)
            return NULL;

        if (head->next == NULL)
            return head;

        ListNode* cur = head;
        ListNode* temp_1 = head->next;
        ListNode* temp_2 = head->next->next;

        cur->next = temp_2;
        temp_1->next = cur;
        head = temp_1;

        ListNode* ahead = cur;
        cur = temp_2;

        while (cur != NULL) {
            if (cur->next == NULL)
                break;

            temp_1 = cur->next;
            temp_2 = cur->next->next;

            ahead->next = temp_1;
            cur->next = temp_2;
            temp_1->next = cur;
            ahead = cur;
            cur = temp_2;
        }

        return head;
    }
};

File: 剑指offer/24.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseList(head *ListNode) *ListNode {
	if head == nil {
		return nil
	}
	res := reverse(head)
	head.Next = nil
	return res
}

func reverse(head *ListNode) *ListNode {
	if head.Next == nil {
		return head
	}
	res := reverse(head.Next)
	head.Next.Next = head
	return res
}
/*
执行结果：通过 显示详情 添加备注
执行用时：0 ms, 在所有 Go 提交中击败了100.00%的用户
内存消耗：3 MB, 在所有 Go 提交中击败了13.21%的用户
*/